aws/resource_aws_msk_cluster.go

```diff
@@ -0,0 +1,307 @@
+}
+func resourceAwsMskClusterRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+	name := d.Get("name").(string)
+
+	state, err := readMskClusterState(conn, name)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("Error reading MSK cluster: \"%s\", code: \"%s\"", awsErr.Message(), awsErr.Code())
+		}
+		return err
+
+	}
+	d.SetId(state.arn)
+	d.Set("arn", state.arn)
+	d.Set("status", state.status)
+	d.Set("encrypt_rest_key", state.encryptRestKey)
+	d.Set("zookeeper_connect", state.zookeeperConnect)
+	d.Set("bootstrap_brokers", state.bootstrapBrokers)
+
+	return nil
+}
+
+func resourceAwsMskClusterDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+	arn := d.Get("arn").(string)
+	name := d.Get("name").(string)
+
+	_, err := conn.DeleteCluster(&kafka.DeleteClusterInput{
+		ClusterArn: aws.String(arn),
+	})
+	if err != nil {
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{kafka.ClusterStateDeleting},
+		Target:     []string{"DESTROYED"},
+		Refresh:    clusterStateRefreshFunc(conn, name),
+		Timeout:    d.Timeout(schema.TimeoutDelete),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for MSK Cluster (%s) to be destroyed: %s",
+			arn, err)
+	}
+
+	return nil
+}
+
+type mskClusterState struct {
+	arn               string
+	creationTimestamp int64
+	status            string
+	encryptRestKey    string
+	zookeeperConnect  string
+	bootstrapBrokers  string
+	brokerCount       int64
+	kafkaVersion      string
+}
+
+func readMskClusterState(conn *kafka.Kafka, name string) (*mskClusterState, error) {
+	listOpts := &kafka.ListClustersInput{
+		ClusterNameFilter: aws.String(name),
+	}
+
+	cluster_list, err := conn.ListClusters(listOpts)
+
+	if len(cluster_list.ClusterInfoList) == 0 {
+		return nil, awserr.New("NotFoundException", fmt.Sprintf("MSK Cluster (%s) not found", name), nil)
+	}
+
+	if len(cluster_list.ClusterInfoList) > 1 {
+		return nil, fmt.Errorf("Ambiguous MSK Cluster name (%s)", name)
+	}
+
+	cluster := cluster_list.ClusterInfoList[0]
+	state := &mskClusterState{}
+
+	if cluster != nil {
+		state.arn = aws.StringValue(cluster.ClusterArn)
+		state.creationTimestamp = aws.TimeValue(cluster.CreationTime).Unix()
+		state.status = aws.StringValue(cluster.State)
+		state.encryptRestKey = aws.StringValue(cluster.EncryptionInfo.EncryptionAtRest.DataVolumeKMSKeyId)
+		state.brokerCount = aws.Int64Value(cluster.NumberOfBrokerNodes)
+		state.zookeeperConnect = aws.StringValue(cluster.ZookeeperConnectString)
+		state.kafkaVersion = aws.StringValue(cluster.CurrentBrokerSoftwareInfo.KafkaVersion)
+
+		if state.status == kafka.ClusterStateActive {
+			bb, bb_err := conn.GetBootstrapBrokers(
```

Comment:
This API call should not be occurring within the `StateChangeConf` refresh function and instead should only be performed in the `Read` function.

Reply:
Duplication is fine. 👍